`SrcLoc` sorting is non-transitive and a false total order and equality (panics in 1.81)

Seems one (or more) of the types used in the transpiler have a wrong `Ord` implementation.
```
Transpiling cr.c
thread 'main' panicked at core/src/slice/sort/shared/smallsort.rs:860:5:
user-provided comparison function does not correctly implement a total order
stack backtrace:
   0:     0x620e4a93b8da - std::backtrace_rs::backtrace::libunwind::trace::h4f532649ff050af0
                               at /rustc/9b72238eb813e9d06e9e9d270168512fbffd7ee7/library/std/src/../../backtrace/src/backtrace/libunwind.rs:116:5
   1:     0x620e4a93b8da - std::backtrace_rs::backtrace::trace_unsynchronized::h2bd9c4760a46d04c
                               at /rustc/9b72238eb813e9d06e9e9d270168512fbffd7ee7/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x620e4a93b8da - std::sys::backtrace::_print_fmt::h06a2f1333d2c9eb7
                               at /rustc/9b72238eb813e9d06e9e9d270168512fbffd7ee7/library/std/src/sys/backtrace.rs:66:9
   3:     0x620e4a93b8da - <std::sys::backtrace::BacktraceLock::print::DisplayBacktrace as core::fmt::Display>::fmt::hc39aee40395be66f
                               at /rustc/9b72238eb813e9d06e9e9d270168512fbffd7ee7/library/std/src/sys/backtrace.rs:39:26
   4:     0x620e4a9628db - core::fmt::rt::Argument::fmt::h391dd44a8ecf182f
                               at /rustc/9b72238eb813e9d06e9e9d270168512fbffd7ee7/library/core/src/fmt/rt.rs:177:76
   5:     0x620e4a9628db - core::fmt::write::h718938895d6a6b80
                               at /rustc/9b72238eb813e9d06e9e9d270168512fbffd7ee7/library/core/src/fmt/mod.rs:1178:21
   6:     0x620e4a937d83 - std::io::Write::write_fmt::h97874dbf951485b0
                               at /rustc/9b72238eb813e9d06e9e9d270168512fbffd7ee7/library/std/src/io/mod.rs:1823:15
   7:     0x620e4a93b722 - std::sys::backtrace::BacktraceLock::print::h28b06058dc60edca
                               at /rustc/9b72238eb813e9d06e9e9d270168512fbffd7ee7/library/std/src/sys/backtrace.rs:42:9
   8:     0x620e4a93c9e7 - std::panicking::default_hook::{{closure}}::ha6790375217144d8
                               at /rustc/9b72238eb813e9d06e9e9d270168512fbffd7ee7/library/std/src/panicking.rs:268:22
   9:     0x620e4a93c816 - std::panicking::default_hook::h7a0db9554a291cc0
                               at /rustc/9b72238eb813e9d06e9e9d270168512fbffd7ee7/library/std/src/panicking.rs:295:9
  10:     0x620e4a93d027 - std::panicking::rust_panic_with_hook::h35232034e4df15db
                               at /rustc/9b72238eb813e9d06e9e9d270168512fbffd7ee7/library/std/src/panicking.rs:801:13
  11:     0x620e4a93ce93 - std::panicking::begin_panic_handler::{{closure}}::hf188e2435ae6fe44
                               at /rustc/9b72238eb813e9d06e9e9d270168512fbffd7ee7/library/std/src/panicking.rs:667:13
  12:     0x620e4a93bdb9 - std::sys::backtrace::__rust_end_short_backtrace::h63f9b1b738507ea6
                               at /rustc/9b72238eb813e9d06e9e9d270168512fbffd7ee7/library/std/src/sys/backtrace.rs:170:18
  13:     0x620e4a93cb54 - rust_begin_unwind
                               at /rustc/9b72238eb813e9d06e9e9d270168512fbffd7ee7/library/std/src/panicking.rs:665:5
  14:     0x620e4a960703 - core::panicking::panic_fmt::ha3d1a45ecd949c55
                               at /rustc/9b72238eb813e9d06e9e9d270168512fbffd7ee7/library/core/src/panicking.rs:74:14
  15:     0x620e4a964d9b - core::slice::sort::shared::smallsort::panic_on_ord_violation::hbf9c589b6c9b3724
                               at /rustc/9b72238eb813e9d06e9e9d270168512fbffd7ee7/library/core/src/slice/sort/shared/smallsort.rs:860:5
  16:     0x620e4a4c7f9f - core::slice::sort::shared::smallsort::bidirectional_merge::h6690bf85da2564d5
                               at /rustc/9b72238eb813e9d06e9e9d270168512fbffd7ee7/library/core/src/slice/sort/shared/smallsort.rs:838:13
  17:     0x620e4a4c7f9f - core::slice::sort::shared::smallsort::small_sort_network::h890c54e4f75c0902
                               at /rustc/9b72238eb813e9d06e9e9d270168512fbffd7ee7/library/core/src/slice/sort/shared/smallsort.rs:370:9
  18:     0x620e4a428107 - <T as core::slice::sort::shared::smallsort::UnstableSmallSortFreezeTypeImpl>::small_sort::hd5067ef6cf7d40f7
                               at /rustc/9b72238eb813e9d06e9e9d270168512fbffd7ee7/library/core/src/slice/sort/shared/smallsort.rs:164:13
  19:     0x620e4a428107 - <T as core::slice::sort::shared::smallsort::UnstableSmallSortTypeImpl>::small_sort::h2df5d3e38e278dc7
                               at /rustc/9b72238eb813e9d06e9e9d270168512fbffd7ee7/library/core/src/slice/sort/shared/smallsort.rs:101:9
  20:     0x620e4a428107 - core::slice::sort::unstable::quicksort::quicksort::hea4648e1605e90a5
                               at /rustc/9b72238eb813e9d06e9e9d270168512fbffd7ee7/library/core/src/slice/sort/unstable/quicksort.rs:24:13
  21:     0x620e4a428db8 - core::slice::sort::unstable::quicksort::quicksort::hea4648e1605e90a5
                               at /rustc/9b72238eb813e9d06e9e9d270168512fbffd7ee7/library/core/src/slice/sort/unstable/quicksort.rs:71:9
  22:     0x620e4a428db8 - core::slice::sort::unstable::quicksort::quicksort::hea4648e1605e90a5
                               at /rustc/9b72238eb813e9d06e9e9d270168512fbffd7ee7/library/core/src/slice/sort/unstable/quicksort.rs:71:9
  23:     0x620e4a512e46 - core::slice::sort::unstable::sort::h35235f49a1d553dd
                               at /rustc/9b72238eb813e9d06e9e9d270168512fbffd7ee7/library/core/src/slice/sort/unstable/mod.rs:43:5
  24:     0x620e4a512e46 - core::slice::<impl [T]>::sort_unstable_by::h504cc4e282cf16fc
                               at /rustc/9b72238eb813e9d06e9e9d270168512fbffd7ee7/library/core/src/slice/mod.rs:2981:9
  25:     0x620e4a512e46 - c2rust_transpile::c_ast::TypedAstContext::sort_top_decls::hc2a55804d3bb0a25
                               at /home/kriskras99/Source/c2rust/c2rust-transpile/src/c_ast/mod.rs:674:9
  26:     0x620e4a512e46 - c2rust_transpile::translator::translate::h599712b2056933b9
                               at /home/kriskras99/Source/c2rust/c2rust-transpile/src/translator/mod.rs:513:9
  27:     0x620e4a5882a5 - c2rust_transpile::transpile_single::he88bad69628bcf0b
                               at /home/kriskras99/Source/c2rust/c2rust-transpile/src/lib.rs:536:9
  28:     0x620e4a3c2fdb - c2rust_transpile::transpile::{{closure}}::h1c26c176d2edae16
                               at /home/kriskras99/Source/c2rust/c2rust-transpile/src/lib.rs:328:17
  29:     0x620e4a3c2fdb - core::iter::adapters::map::map_fold::{{closure}}::he4bccb3109727379
                               at /rustc/9b72238eb813e9d06e9e9d270168512fbffd7ee7/library/core/src/iter/adapters/map.rs:88:28
  30:     0x620e4a3c2fdb - <core::slice::iter::Iter<T> as core::iter::traits::iterator::Iterator>::fold::hd20ded494caaffcb
                               at /rustc/9b72238eb813e9d06e9e9d270168512fbffd7ee7/library/core/src/slice/iter/macros.rs:232:27
  31:     0x620e4a3c2fdb - <core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::fold::h45f09cacc691c60d
                               at /rustc/9b72238eb813e9d06e9e9d270168512fbffd7ee7/library/core/src/iter/adapters/map.rs:128:9
  32:     0x620e4a3c2fdb - core::iter::traits::iterator::Iterator::for_each::h2c40a859fa37c766
                               at /rustc/9b72238eb813e9d06e9e9d270168512fbffd7ee7/library/core/src/iter/traits/iterator.rs:813:9
  33:     0x620e4a3c2fdb - alloc::vec::Vec<T,A>::extend_trusted::hb59eb0450c1a5309
                               at /rustc/9b72238eb813e9d06e9e9d270168512fbffd7ee7/library/alloc/src/vec/mod.rs:3125:17
  34:     0x620e4a3c2fdb - <alloc::vec::Vec<T,A> as alloc::vec::spec_extend::SpecExtend<T,I>>::spec_extend::hdfb895ec3556694f
                               at /rustc/9b72238eb813e9d06e9e9d270168512fbffd7ee7/library/alloc/src/vec/spec_extend.rs:26:9
  35:     0x620e4a3c2fdb - <alloc::vec::Vec<T> as alloc::vec::spec_from_iter_nested::SpecFromIterNested<T,I>>::from_iter::ha75a8e9b8b1dfed4
                               at /rustc/9b72238eb813e9d06e9e9d270168512fbffd7ee7/library/alloc/src/vec/spec_from_iter_nested.rs:60:9
  36:     0x620e4a3c2fdb - <alloc::vec::Vec<T> as alloc::vec::spec_from_iter::SpecFromIter<T,I>>::from_iter::hd503974f9b2fb62a
                               at /rustc/9b72238eb813e9d06e9e9d270168512fbffd7ee7/library/alloc/src/vec/spec_from_iter.rs:33:9
  37:     0x620e4a584ece - <alloc::vec::Vec<T> as core::iter::traits::collect::FromIterator<T>>::from_iter::hdcb5e575103de682
                               at /rustc/9b72238eb813e9d06e9e9d270168512fbffd7ee7/library/alloc/src/vec/mod.rs:2989:9
  38:     0x620e4a584ece - core::iter::traits::iterator::Iterator::collect::h748ec39779ff7b21
                               at /rustc/9b72238eb813e9d06e9e9d270168512fbffd7ee7/library/core/src/iter/traits/iterator.rs:2000:9
  39:     0x620e4a584ece - c2rust_transpile::transpile::hd0035b677d692a90
                               at /home/kriskras99/Source/c2rust/c2rust-transpile/src/lib.rs:337:14
  40:     0x620e4a35911d - c2rust_transpile::main::h74e6ff961002307d
                               at /home/kriskras99/Source/c2rust/c2rust/src/bin/c2rust-transpile.rs:259:5
  41:     0x620e4a361a33 - core::ops::function::FnOnce::call_once::h5ca91e7049f14308
                               at /rustc/9b72238eb813e9d06e9e9d270168512fbffd7ee7/library/core/src/ops/function.rs:250:5
  42:     0x620e4a361a33 - std::sys::backtrace::__rust_begin_short_backtrace::hd38ba6c044c85727
                               at /rustc/9b72238eb813e9d06e9e9d270168512fbffd7ee7/library/std/src/sys/backtrace.rs:154:18
  43:     0x620e4a361a29 - std::rt::lang_start::{{closure}}::h756998bcb6544777
                               at /rustc/9b72238eb813e9d06e9e9d270168512fbffd7ee7/library/std/src/rt.rs:164:18
  44:     0x620e4a933200 - core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once::h5412c97e5e68829b
                               at /rustc/9b72238eb813e9d06e9e9d270168512fbffd7ee7/library/core/src/ops/function.rs:284:13
  45:     0x620e4a933200 - std::panicking::try::do_call::h3d002e8ffa4f91b7
                               at /rustc/9b72238eb813e9d06e9e9d270168512fbffd7ee7/library/std/src/panicking.rs:557:40
  46:     0x620e4a933200 - std::panicking::try::h60529ec65ad686dc
                               at /rustc/9b72238eb813e9d06e9e9d270168512fbffd7ee7/library/std/src/panicking.rs:520:19
  47:     0x620e4a933200 - std::panic::catch_unwind::h28f4a63bd2319cba
                               at /rustc/9b72238eb813e9d06e9e9d270168512fbffd7ee7/library/std/src/panic.rs:345:14
  48:     0x620e4a933200 - std::rt::lang_start_internal::{{closure}}::h14b8d458d7b543a7
                               at /rustc/9b72238eb813e9d06e9e9d270168512fbffd7ee7/library/std/src/rt.rs:143:48
  49:     0x620e4a933200 - std::panicking::try::do_call::h9747ac76d396a084
                               at /rustc/9b72238eb813e9d06e9e9d270168512fbffd7ee7/library/std/src/panicking.rs:557:40
  50:     0x620e4a933200 - std::panicking::try::h0f984b4d1c23e9bd
                               at /rustc/9b72238eb813e9d06e9e9d270168512fbffd7ee7/library/std/src/panicking.rs:520:19
  51:     0x620e4a933200 - std::panic::catch_unwind::hcd3e02aab11d9c76
                               at /rustc/9b72238eb813e9d06e9e9d270168512fbffd7ee7/library/std/src/panic.rs:345:14
  52:     0x620e4a933200 - std::rt::lang_start_internal::h370157592acdbc7d
                               at /rustc/9b72238eb813e9d06e9e9d270168512fbffd7ee7/library/std/src/rt.rs:143:20
  53:     0x620e4a35dbac - main
  54:     0x787e58034e08 - <unknown>
  55:     0x787e58034ecc - __libc_start_main
  56:     0x620e4a354325 - _start
  57:                0x0 - <unknown>
```

To reproduce:
```
cargo +1.81.0 install --git https://github.com/immunant/c2rust.git c2rust
git clone https://github.com/chirlu/soxr
cd soxr
cmake -S . -B build -DCMAKE_BUILD_TYPE=None -DCMAKE_INSTALL_PREFIX=/usr -DCMAKE_EXPORT_COMPILE_COMMANDS=ON -DBUILD_EXAMPLES=OFF -DBUILD_SHARED_LIBS=ON -DWITH_AVFFT=ON -DWITH_LSR_BINDINGS=ON -DWITH_OPENMP=OFF -DWITH_PFFFT=ON
c2rust transpile -o rust build/compile_commands.json
```

Thanks for the hint! Unfortunately I'm now getting this failure trying to build it like that.
```
  Cloning into 'tinycbor_build'...
  HEAD is now at d393c16 Parser: fix reading it->extra on big endian when bytesNeeded == 1
  make[4]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
  make[4]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
  make[4]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
  make[4]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
  src/ExportResult.hpp:11:10: fatal error: 'algorithm' file not found
  Unable to generate ExportResult bindings
````
Now I've updated to the Pop_OS 24.04 release in the meantime so I'm not sure if I've run into something due to that upgrade or whether I've just cloned it at an unfortunate time?

--- a/c2rust-transpile/src/c_ast/mod.rs
+++ b/c2rust-transpile/src/c_ast/mod.rs
@@ -155,13 +155,17 @@ struct SrcLocInclude<'a> {
     include_path: &'a [SrcLoc],
 }
 
+impl SrcLocInclude<'_> {
+    fn cmp_iter<'a>(&'a self) -> impl Iterator<Item = SrcLoc> + 'a {
+        // See docs on `Self` for why this is the right comparison.
+        let Self { loc, include_path } = *self;
+        include_path.iter().copied().chain([loc])
+    }
+}
+
 impl PartialOrd for SrcLocInclude<'_> {
     fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
-        let a = self;
-        let b = other;
-        let a = a.include_path.iter().copied().chain([a.loc]);
-        let b = b.include_path.iter().copied().chain([b.loc]);
-        Some(a.cmp(b))
+        Some(self.cmp_iter().cmp(other.cmp_iter()))
     }
 }
 

